docs/configuration.md
# Configuring Turnilo

It is easy to start using Turnilo with Druid by pointing it at your Druid cluster: `turnilo --druid broker_host:broker_port`
Turnilo will automatically introspect your Druid cluster and figure out available datasets.

Turnilo can be configured with a *config* YAML file. While you could write one from scratch it is recommended to let
Turnilo give you a head start by using it to generate a config file for you using the default introspection.

Run:

```bash
turnilo --druid broker_host:broker_port --print-config --with-comments > config.yaml
```

This will cause Turnilo to go through its normal startup and introspection routine and then dump the internally generated
config (complete with comments) into the provided file.

You can now run `turnilo --config config.yaml` to run Turnilo with your config.

The next step is to open the generated config file in your favourite text editor and configure Turnilo to your liking.
Below we will go through a typical configuration flow. At any point you can save the config and re-launch Turnilo to load
that config in.


## Configuring the Turnilo server

**port** (number), default: 9090

The port that Turnilo should run on.

**verbose** (boolean), default: false

Indicates that Turnilo should run in verbose mode. This will log all the queries done by Turnilo.

**serverHost** (string), default: bind to all hosts

The host that Turnilo will bind to.

**serverRoot** (string), default: ""

A custom path to act as the server string.

The Turnilo UI will be served from `http://turnilo-host:$port/$serverRoot`

**serverTimeout** (number), default: 0

Timeout on all server request handlers in ms. Default value is 0 which means no timeout. 
Turnilo sets timeout on Response object for every incoming request. If response is not send before timeout, Turnilo closes connection.

**clientTimeout** (number), default: 0

Timeout on client requests in ms. Default value is 0 which means no timeout. 
Timeout is set in browser on every request to Turnilo server.

**readinessEndpoint** (string), default "/health/ready"

Readiness endpoint location. Checks readiness of druid clusters. See [Checking health of Turnilo instance](health-checking.md)

**livenessEndpoint** (string), default "/health/alive"

Liveness endpoint location. Reports liveness status of turnilo app.

**healthEndpoint** (string), deprecated

Old value that will be used as readinessEndpoint.

**iframe** ("allow" | "deny"), default "allow"

Specify whether Turnilo will be allowed to run in an iFrame.
If set to "deny" Turnilo will set the following headers:

```
X-Frame-Options: "DENY"
Content-Security-Policy: "frame-ancestors 'none'"
```

This is used to prevent [Clickjacking](http://en.wikipedia.org/wiki/clickjacking).
Learn more about it on [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Frame-Options).

**trustProxy** ("none" | "always"), default "none"

Should the server trust the `X-Forwarded-*` headers.  If "always", Turnilo will use the left-most entry from the header.

**strictTransportSecurity** ("none" | "always"), default "none"

Specify that Turnilo should set the [StrictTransportSecurity](https://developer.mozilla.org/en-US/docs/Web/Security/HTTP_strict_transport_security) header.

Note that Turnilo can itself only run an http server.
This option is intended to be used when when Turnilo is running behind a HTTPS terminator like AWS ELB.


## Configuring the Clusters

The top level `clusters:` key that holds the clusters that Turnilo will connect to.


### General properties

Each cluster has the following properties:

**name** (string)

The name of the cluster (to be referenced later from the data cube).

**type** ('druid')

The database type of the cluster. Currently only Druid is supported.

**url** (string)

The url address (http[s]://hostname[:port]) of the cluster. If no port, 80 is assumed for plain http, and 443 for secure https.

**host** deprecated (string)

The host (hostname:port) of the cluster, http protocol is assumed. Deprecated, use **url** field

**version** (string)

The explicit version to use for this cluster.
Define this to override the automatic version detection.

**timeout** (number)

The timeout to set on the Druid queries in ms. See [documentation](https://druid.apache.org/docs/latest/querying/query-context.html)

**retry** (object)

Options for retries on Druid native queries. If no object is provided Turnilo will not retry failed queries.
Object should have following structure:

```yaml
retry:
    maxAttempts: 10
    delay: 1000
``` 

* `maxAttempts` is count of maximum attempts for retry. Default values is 5
* `delay` is time in ms between each attempt.

**healthCheckTimeout** (number), default: 1000

The timeout for the cluster health checking request in ms. See [Checking health of Turnilo instance](health-checking.md)

**sourceListScan** ("auto" | "disable"), default: "auto"

Should the sources of this cluster be automatically scanned and new sources added as data cubes.

**sourceListRefreshOnLoad** (boolean), default: false

Should the list of sources be reloaded every time that Turnilo is loaded.
This will put additional load on the data store but will ensure that sources are visible in the UI as soon as they are created.

**sourceListRefreshInterval** (number), minimum: 1000, default: 0

How often should sources be reloaded in ms. Default value of 0 disables periodical source refresh.

**sourceReintrospectOnLoad** (boolean), default: false

Should sources be scanned for additional dimensions every time that Turnilo is loaded.
This will put additional load on the data store but will ensure that dimension are visible in the UI as soon as they are created.

**sourceReintrospectInterval** (number), minimum: 1000, default: 0

How often should source schema be reloaded in ms. Default value of 0 disables periodical source refresh.


### Druid specific properties

**introspectionStrategy** ("segment-metadata-fallback" | "segment-metadata-only" | "datasource-get"), default: "segment-metadata-fallback"

The introspection strategy for Druid cluster.

**requestDecorator** (string)

The request decorator module filepath to load.

**decoratorOptions** (any)

Options passed to the request decorator module


## Configuring Data Cubes

The top level `dataCubes:` key that holds the data cubes that will be loaded into Turnilo.
The order of the data cubes in the config will define the ordering seen in the UI.


### Basic data cube properties

Described here are only the properties which you might want to change.

**name** (string)

The name of the data cube as used internally in Turnilo and used in the URLs. This should be a URL safe string.
Changing this property for a given data cube will break any URLs that someone might have generated for that data
cube in the past.

**title** (string)

The user visible name that will be used to describe this data cube in the UI. It is always safe to change this.

**description** (markdown)

The description of the data cube in markdown format. Description is shown on home page. 
If description contains horizontal line (markdown: ` --- `) it will split description and 
later part will be visible after clicking "Show more" button on UI.

**clusterName** (string)

The cluster that the data cube belongs to (or "native" if this is a file based data cube).

**source** (string or string[])

Druid data source name or names. Multiple names result in [Union Data Source](http://druid.io/docs/latest/querying/datasource.html)

**defaultTimezone** (string - timezone)

The default timezone, expressed as an [Olsen Timezone](https://en.wikipedia.org/wiki/Tz_database),
that will be selected when the user first opens this cube. Default `Etc/UTC`.

**defaultDuration** (string - duration), default P1D (1 day)

The time period, expressed as an [ISO 8601 Duration](https://en.wikipedia.org/wiki/ISO_8601#Durations),
that will be shown when the user first opens this cube.

**defaultSortMeasure** (string), default: the first measure

The name of the measure that will be used for default sorting. It is commonly set to the measure that represents the
count of events.

**defaultSelectedMeasures** (string[]), default: first four measures

The names of the measures that will be selected by default.

**defaultPinnedDimensions** (string[])

The names of the dimensions (in order) that will appear *pinned* by default on the right panel.

**introspection** ("none" | "no-autofill" | "autofill-dimensions-only" | "autofill-measures-only" | "autofill-all")

Data cube introspection strategy.

**subsetFormula** (string - plywood expression)

A filter defined as [Plywood expression](http://plywood.imply.io/expressions) that is applied to data cube. This expression is not shown in the filter area, so Turnilo user is not aware of subset formula filter clause.

**refreshRule**

Refresh rule defining how the information about latest data in a data source is obtained.

**maxSplits** (number)

Number of splits data cube supports. Defaults to 3.

**maxQueries** (number)

Number of queries that can be issued to druid. Defaults to 500.

### Refresh rules

The `refreshRule:` section of the data cube allows the customisation of latest data discovery mechanism.

**rule** ("query" | "realtime" | "fixed" ), default: "query"

The name of the rule which will be used to obtain information about the latest data. Following rules are available:

- `query`: best suited for batch data sources. The data source will be queried every minute to obtain the maximum value from time dimension.
- `realtime`: best suited for realtime data sources. The data source will not be queried and the value of *now* is assumed as a latest data time.
- `fixed`: best suited for constant data sources. The data source will not be queried and the value of `refreshRule.time` property will be used.

**time** (string - date with time instant)

Latest date time of a data source, expressed as [ISO 8601 Instant](https://en.wikipedia.org/wiki/ISO_8601#Combined_date_and_time_representations).
Applicable only if `refreshRule.rule` is set to `fixed`.

### Attribute Overrides

While Turnilo tries to learn as much as it can from your data cube from Druid directly.
It can not (yet) do a perfect job. The `attributeOverrides:` section of the data cube is there for you to fix that.

**name** (string)

The name of the attribute (column) in Druid. This must match the Druid name.

Here are some common scenarios where you should add an attribute override:

#### Override native Druid type

If Turnilo is not able to discover column type, it could be overridden with Druid native column type. 

```yaml
- name: column_name
  nativeType: hyperUnique
```

Supported native types: "hyperUnique", "thetaSketch" and "approximateHistogram".

#### Override numeric dimension

Turnilo could not correctly detect numeric dimensions as Druid reports all dimensions to be strings.
When a numeric dimension is incorrectly classified as a string its sorting will appear wrong in the UI.
If you have a dimension with numeric values (say: `age`).

You should add:

```yaml
- name: age
  type: NUMBER
```

To the `attributeOverrides` to tell Turnilo that this is numeric.

You can now use `$age` in numeric expressions. For example you could create a dimension with the formula
`$age / 2 + 7`.


### Dimensions

In this section you can define the dimensions that users can *split* and *filter* on in the UI.
Dimensions may be organized as list or tree where each item of list can be either a dimension
or [dimension group](#dimension-group) having its own dimensions list or tree.
The order of the dimension list in the top of the left panel is determined by the order of the dimensions definitions
in this section.


#### Dimension

Dimensions are defined with following attributes:

**name** (string)

The name of the dimension.
This does not have to correspond to the attribute name (but the auto generated dimensions do).
This should be a URL safe string and unique across dimensions, dimension groups, measures and measure groups.
Changing this property will break any URLs that someone might have generated that include this dimension.

**title** (string)

The title for this dimension in the UI. Can be anything and is safe to change at any time.

**description** (string)

The description of the dimension in the UI. Accepts Markdown format.

**url** (string)

A url associated with the dimension, with optional token '%s' that is replaced by the dimension value to generate
a link specific to each value.

**granularities** (string[5] or number[5]), default: ["PT1M", "PT5M", "PT1H", "P1D", "P1W"]`

For time dimensions you can define a set of exactly 5 granularities that you want to be available for bucketing.

Each granularity must be expressed as a 'floorable' [ISO 8601 duration](https://en.wikipedia.org/wiki/ISO_8601#Durations)
A floorable duration is any duration that is either a single period like `P1D`, `P1W`, `PT1S`, e.t.c. or a multiple period
that divides into the larger period. For example, `P3M` is floorable because 3 divides 12 but `P5M` is not floorable.

If you mainly care about smaller intervals, you might want to set it to: `['PT1S', 'PT15S', 'PT30S', 'PT1M', 'PT1H']`

Alternatively, if you mainly care about large intervals, you might want to try: `['P1D', 'P1W', 'P1M', 'P3M', 'P1Y']`

For number dimensions you can just provide 5 bucket sizes as integers.

**bucketingStrategy** ("defaultBucket" | "defaultNoBucket")

Specify whether or not the dimension should be bucketed by default. If unspecified defaults to 'defaultBucket' for time and numeric dimensions.

**sortStrategy** ("self" | `someMeasureName`)

Specify a specific sort strategy for this dimension in visualizations. If unspecified defaults to best sort strategy based on the visualization.

**kind** ("string" | "boolean" | "number" | "time")

Specify kind of data inside dimension. It defaults to "string".

**multiValue** (boolean)

Set to true if dimension holds multiple values. [Druid Multi-Value Dimensions](https://druid.apache.org/docs/latest/querying/multi-value-dimensions.html)

**formula** (string - plywood expression)

The [Plywood expression](http://plywood.imply.io/expressions) for this dimension.
By default it is `$name` where *name* is the name of the dimension.
You can create derived dimensions by using non-trivial formulas.

Here are some common use cases for derived dimensions:

##### Lookup formula

If you have a dimension that represents an ID that is a key into some other table. You may have set up a
[Druid Query Time Lookup](http://druid.io/docs/latest/querying/lookups.html) in which case you could

```yaml
- name: correctValue
  formula: $lookupKey.lookup('my_awesome_lookup')
```

Which would apply the lookup.

You can also apply the `.fallback()` action as ether:

- `$lookupKey.lookup('my_awesome_lookup').fallback($lookupKey)` to keep values that were not found as they are.
- `$lookupKey.lookup('my_awesome_lookup').fallback('missing')` to map missing values to the word 'missing'.

##### Extraction formula

Imagine you have an attribute `resourceName` which has values:

```json
["druid-0.8.2", "druid-0.8.1", "druid-0.7.0", "index.html"]
```

You could apply, for example, the `.extract` function by creating the following dimension:

```yaml
- name: resourceVersion
  formula: $resourceName.extract('(\d+\.\d+\.\d+)')
```

Which would have values:

```json
["0.8.2", "0.8.1", "0.7.0", null]
```

##### Boolean formula

It is often useful to create dimensions that are the result of some boolean expression.
Let's say that you are responsible for all accounts in the United States as well as some specific account you could create a dimension like:

```yaml
- name: myAccounts
  formula: $country == 'United States' or $accountName.in(['Toyota', 'Honda'])
```

Now my account would represent a custom filter boolean dimension.

##### Quantiles

If you have dimension defined as histogram, you can add quantile measure. Use plywood method quantile on desired histogram and provide required parameters.
Percentile parameter would be used as default percentile and could be adjusted on UI. Tunning parameters will be passed as is to Druid. 

```yaml
- name: clicks_percentile
  formula: $main.quantile($response_time_ms, 0.99, 'k=128')
```

Turnilo can handle percentiles only as top level operation in expression so it is impossible to nest quantile expression inside let's say division. 

```yaml
- name: opaque_percentile_formula
  forumla: $main.quantile($response_time_ms, 0.9, 'k=128') * 1000
```

If turnilo encounters such formula, it would assume it is simple measure. User would be able to use this measure as is, but won't be able to picking percentile. 

##### Custom transformations

If no existing plywood function meets your needs, you could also define your own custom transformation.
The transformation could be any supported [Druid extraction function](http://druid.io/docs/latest/querying/dimensionspecs.html).

For example you could apply any number of javascript functions to a string.

To use that in Turnilo define following `options` at data cube level:

```yaml
options:
  customTransforms:
    stringFun:
      extractionFn:
        type: javascript
        function: function(x) { try { return decodeURIComponent(x).trim().charCodeAt(0) } catch(e) { return null; } }
```

Then in the dimensions simply reference `stringFun` like so:

```yaml
- name: stringFun
  title: String Fun
  formula: $countryURL.customTransform('stringFun')
```

#### Dimension Group

Dimension groups are defined with following attributes:

**name** (string)

The name of the dimension group.
This should be a URL safe string and unique across dimensions, dimension groups, measures and measure groups.

**title** (string)

The title for this dimension group in the UI. Can be anything and is safe to change at any time.

**description** (string)

The description of the dimension group in the UI. Accepts Markdown format.

**dimensions** (Dimension | DimensionGroup)[]

An array of nested dimensions or dimension groups. It cannot be empty.

### Measures

In this section you can define the measures that users can *aggregate* on (*apply*) on in the UI.
Measures may be organized as list or tree where each item of list can be either a measure
or [measure group](#measure-group) having its own measures list or tree.
The order of the measure list in the bottom of the left panel is determined by the order of the measure definitions
in this section.

#### Measure

Measures are defined with following attributes:

**name** (string)

The name of the measure.
This should be a URL safe string.
Changing this property will break any URLs that someone might have generated that include this measure.

**title** (string)

The title for this measure in the UI. Can be anything and is safe to change at any time.

**description** (string)

The description of the measure in the UI. Accepts Markdown format.

**units** (string)

The units for this measure. To be shown alongside the title.

**lowerIsBetter** (boolean) default: false

If set to true measure will be interpreted in UI as "better" when value is lower.

**format** (string)  default: "0,0.0 a"

Default format for measure as string in [numbro format](http://numbrojs.com/old-format.html).

**formula** (string - plywood expression)

The [Plywood expression](http://plywood.imply.io/expressions) for this dimension. By default it is `$main.sum($name)` where *name* is the name of the measure.

The `$main` part of the measure expressions serves as a placeholder for the data segment.
In Plywood every aggregate is a function that acts on a data segment.

**transformation** ("none", "percent-of-parent", "percent-of-total"), default: "none"

Predefined transformation that can be applied to a measure formula. Currently supported options are:
- "none" - no transformation;
- "percent-of-parent" - displays the ratio between expression value at current split and expression value at parent split;
- "percent-of-total" - displays the ratio between expression value at current split and total expression value without splitting.


One can also create derived measures by using non-trivial expressions in **formula**. Here are some common use cases for derived dimensions:


##### Ratio formula

Ratios are generally considered fun.

```yaml
- name: ecpm
  title: eCPM
  formula: $main.sum($revenue) / $main.sum($impressions) * 1000
```


##### Filtered aggregations formula

A very powerful tool is to use a filtered aggregate.
If, for example, your revenue in the US is a very important measure you could express it as:

```yaml
- name: usa_revenue
  title: USA Revenue
  formula: $main.filter($country == 'United States').sum($revenue)
```

It is also common to express a ratio of something filtered vs unfiltered.

```yaml
- name: errorRate
  formula: $main.filter($statusCode == 500).sum($requests) / $main.sum($requests)
```


##### Custom aggregations

Within the measures you have access to the full power of the [Plywood expressions](http://plywood.imply.io/expressions).
If you ever find yourself needing to go beyond the expressive potential of Plywood you could define your own custom aggregation.
The aggregation could be any supported Druid aggregation.

For example Plywood currently does not support the modulo operator.
While Druid has no native modulo support ether it is possible to modulo a measure by using a [javascript aggregator](http://druid.io/docs/latest/querying/aggregations.html#javascript-aggregator).


CHANGELOG.md
# Change Log

## 0.9.39

- Fixed numeric dimension extent computation

## 0.9.38

- Fixed side drawer animation

## 0.9.37

- Dump any and all nodejitsu dependencies added to shrinkwrap
- Contains and Regexp search filters

## 0.9.36

- Added ability to define custom extraction functions
- Added ability to define custom druid context

## 0.9.35

- Fixed unit forwarding

## 0.9.34

- Added ability to customize `logoutHref`

## 0.9.33

- Fix CSS for small screens

## 0.9.32

- Add `--server-host` command line var
- Add `--server-root` command line var

## 0.9.31

- Fix `defaultSplits` in `dataCube` 

## 0.9.30

- New plywood + more back compat tests

## 0.9.29

- Added user_email to request tracker

## 0.9.28

- Added basic request tracking ability
- Changed default request log format and made it configurable

## 0.9.27

- Added gear icon links to settings

## 0.9.26

- DataCube `subsetFilter` -> `subsetFormula`
- Settings fixes

## 0.9.25

- Settings progress

## 0.9.24

- Added trustProxy to server options
- Removed 'when-x-forwarded-proto'

## 0.9.23

- Adding strictTransportSecurity to server options
- Changing ServerSettings to follow new defaults pattern

## 0.9.22

- Moved default user

## 0.9.21

- Renamed `DataSource` to `DataCube`
- Changed Dimension and Measure `expression` to `formula`

## 0.9.20

- Fixes for sorting on dimension when there are multiple splits

## 0.9.19

- Fixes for data source transitions

## 0.9.18

- `DataSource#engine` renamed to `DataSource#clusterName`
- Cleaned up integration tests

## 0.9.17

- Added better small screen support
- Fixed dragging polyfill (on touch devices)
- Added file notices

## 0.9.16

- Selected filtered items appear at top of menu

## 0.9.15

- Allow continuous dimension in two split line chart
- User menu everywhere

## 0.9.14

- New home view
- Added descriptions to data sources

## 0.9.13

- New plywood with fix for: https://github.com/implydata/plywood/pull/121

## 0.9.12

- New plywood fixes bugs

## 0.9.11

- Do not shrinkwrap npm modules

## 0.9.10

- Fix sort ordering of granularities in the granularity picker
- Single measure mode will re-sort on click
- Configurable timezones

## 0.9.9

- Auto generated config links to a github tag
- Fix config auto-generator printing out `defaultSelectedMeasures: []` 

## 0.9.8

- Early detection of (invalid) data sources that have a measure and dimension of the same name
- Prevention of making data sources with measures and dimensions of the same name 

## 0.9.7

- Tiny style fix

## 0.9.6

- Fix sourceListScan introspection to introspect unknown sources as they appear

## 0.9.5

- Fixed display of exclusion filters 
- Ensure cluster connection prior to starting introspection

## 0.9.4

- Exclusion filters

## 0.9.3

- Removed tests, docs, resources, e.t.c from npm module

## 0.9.2

- Change sourceListScan back to default to auto
- Typos and misc error message improvements

## 0.9.1

- Complete refactor of settings architecture
- Settings can be updated at runtime
- Swiv can connect to multiple clusters
- Swiv can connect to Postgres and MySQL
- Full support of Druid 0.9.1
- Auth module versioning
- Command line flag changes (see [writeup](docs/swiv-0.9.x-migration.md))
- Druid Request decorator changes (see [writeup](docs/swiv-0.9.x-migration.md))
- Ability to [generate programmatic links](docs/generating-links-into-swiv.md) into Swiv
- Ability to configure `serverHost`
- Auto refresh icon rotation
- Swiv is npm shrinkwrapped
- Better dimension / measure panel scaling 

## 0.8.42

- Fixed bug in Print config and added CLI tests

## 0.8.41

- New overflow style in split bar also
- Ability to define defaultSelectedMeasures
- Misc fixes for time filter rendering

## 0.8.40

- New Plywood adds support for native quantiles
- New filter menu overflow style

## 0.8.39

- Fixed example config

## 0.8.38

- Continuous dimension filtering and splitting

## 0.8.37

- Fixed small bar bug

## 0.8.36

- Misc query improvements from new Plywood 

## 0.8.35

- Two splits with scrolling in bar chart
- Swiv will only query within server root

## 0.8.34

- New table scroller
- About dialog shows version

## 0.8.33

- Fixed error when going from Time Series to Bar Chart

## 0.8.32

- Configurable server root (default: `/swiv`)

## 0.8.31

- Support and auto detection for Theta sketches (for countDistinct() only for now) 

## 0.8.30

- Horizontal bars in Table
- Misc visualization selection fixes

## 0.8.29

- Side panel resizing

## 0.8.28

- Visualization class refactor
- Table highlight bubble now shows full value
- Better word wrapping in highlight bubbles
- New Plywood ensures to send UTC timestamps to Druid
- Ability to define custom granularities for bucketing

## 0.8.27

- New chronoshift fixes bug with flooring `PT12H`
- Fix burger menu overflow

## 0.8.26

- Better time axis for tiny intervals

## 0.8.25

- New Plywood (0.10.14)

## 0.8.24

- Improved performance of raw data modal
- Added all US timezones

## 0.8.23

- Fixed npm publish issue

## 0.8.22

- Timezone support
- Date range picker
- Fixed X-axis labeling

## 0.8.21

- Build system works with node 6
- Ability to set custom page title

## 0.8.20

- Added pre-calculation pattern to visualizations
- Fixed bar chart on negative values 

## 0.8.19

- More robust cross browser dragging
- Faster incremental build system

## 0.8.18

- Better favicon
- Fixed natural bar chart order

## 0.8.17

- Build `gulp` will fail (exit 1) on error

## 0.8.16

- Fixed sort metric when following auto suggestion

## 0.8.15

- Export data to CSV
- New time series bubbles with hover logic
- Raw data modal allows you to see the raw data in the selected segment
- Better automatic config generation for data sources with URL unsafe names 

## 0.8.14

- Published due to script error (unpublished, please ignore)

## 0.8.13

- Published due to script error (unpublished, please ignore)

## 0.8.12


src/server/config.ts
/*
 * Copyright 2015-2016 Imply Data, Inc.
 * Copyright 2017-2019 Allegro.pl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as nopt from "nopt";
import * as path from "path";
import { LOGGER, NULL_LOGGER } from "../common/logger/logger";
import { AppSettings } from "../common/models/app-settings/app-settings";
import { Cluster } from "../common/models/cluster/cluster";
import { DataCube } from "../common/models/data-cube/data-cube";
import { arraySum } from "../common/utils/general/general";
import { appSettingsToYAML } from "../common/utils/yaml-helper/yaml-helper";
import { ServerSettings, ServerSettingsJS } from "./models/server-settings/server-settings";
import { loadFileSync } from "./utils/file/file";
import { SettingsManager } from "./utils/settings-manager/settings-manager";
import { SettingsStore } from "./utils/settings-store/settings-store";

const AUTH_MODULE_VERSION = 1;
const PACKAGE_FILE = path.join(__dirname, "../../package.json");

function exitWithMessage(message: string): void {
  console.log(message);

  // Hack: load the package file for no reason other than to make some time for console.log to flush
  try {
    loadFileSync(PACKAGE_FILE, "json");
  } catch (e) {
  }

  process.exit();
}

function exitWithError(message: string): void {
  console.error(message);
  process.exit(1);
}

function zeroOne(thing: any): number {
  return Number(Boolean(thing));
}

var packageObj: any = null;
try {
  packageObj = loadFileSync(PACKAGE_FILE, "json");
} catch (e) {
  exitWithError(`Could not read package.json: ${e.message}`);
}
export const VERSION = packageObj.version;

const USAGE = `
Usage: turnilo [options]

Possible usage:

  turnilo --examples
  turnilo --druid http://your.broker.host:8082

General arguments:

      --help                   Print this help message
      --version                Display the version number
  -v, --verbose                Display the DB queries that are being made

Server arguments:

  -p, --port <port-number>     The port turnilo will run on (default: ${ServerSettings.DEFAULT_PORT})
      --server-host <host>     The host on which to listen on (default: all hosts)
      --server-root <root>     A custom server root to listen on (default ${ServerSettings.DEFAULT_SERVER_ROOT})

Data connection options:

  Exactly one data connection option must be provided.

  -c, --config <path>          Use this local configuration (YAML) file
      --examples               Start Turnilo with some example data for testing / demo
  -f, --file <path>            Start Turnilo on top of this file based data cube (must be JSON, CSV, or TSV)
  -d, --druid <url>            The url address (http[s]://hostname[:port]) of the druid broker. If no port, 80 is assumed for plain http, and 443 for secure https.

Configuration printing utilities:

      --print-config           Prints out the auto generated config and exits
      --with-comments          Adds comments when printing the auto generated config
`;

function parseArgs() {
  return nopt(
    {
      "help": Boolean,
      "version": Boolean,
      "verbose": Boolean,

      "port": Number,
      "server-host": String,
      "server-root": String,

      "examples": Boolean,
      "example": String, // deprecated
      "config": String,
      "auth": String,

      "print-config": Boolean,
      "with-comments": Boolean,

      "file": String,
      "druid": String
    },
    {
      v: ["--verbose"],
      p: ["--port"],
      c: ["--config"],
      f: ["--file"],
      d: ["--druid"]
    },
    process.argv
  );
}

var parsedArgs = parseArgs();

if (parsedArgs["help"]) {
  exitWithMessage(USAGE);
}

if (parsedArgs["version"]) {
  exitWithMessage(VERSION);
}

if (parsedArgs["example"]) {
  delete parsedArgs["example"];
  parsedArgs["examples"] = true;
}

const SETTINGS_INPUTS = ["config", "examples", "file", "druid", "postgres", "mysql"];

var numSettingsInputs = arraySum(SETTINGS_INPUTS.map(input => zeroOne(parsedArgs[input])));

if (numSettingsInputs === 0) {
  exitWithMessage(USAGE);
}

if (numSettingsInputs > 1) {
  console.error(`only one of --${SETTINGS_INPUTS.join(", --")} can be given on the command line`);
  if (parsedArgs["druid"] && parsedArgs["config"]) {
    console.error("Looks like you are using --config and --druid in conjunction with each other");
    console.error("This usage is no longer supported. If you are migrating from Swiv < 0.9.x");
    console.error("Please visit: (https://github.com/yahoo/swiv/blob/master/docs/swiv-0.9.x-migration.md)");
  }
  process.exit(1);
}

export const PRINT_CONFIG = Boolean(parsedArgs["print-config"]);
export const START_SERVER = !PRINT_CONFIG;
const logger = START_SERVER ? LOGGER : NULL_LOGGER;

// Load server settings
let configPath = parsedArgs["config"];

if (parsedArgs["examples"]) {
  configPath = path.join(__dirname, "../../config-examples.yaml");
}

let serverSettingsJS: ServerSettingsJS;
let configDirPath;
let configContent;
if (configPath) {
  configDirPath = path.dirname(configPath);
  try {
    configContent = loadFileSync(configPath, "yaml");
    serverSettingsJS = configContent;
    logger.log(`Using config ${configPath}`);
  } catch (e) {
    exitWithError(`Could not load config from '${configPath}': ${e.message}`);
  }
} else {
  configDirPath = process.cwd();
  serverSettingsJS = {};
}

if (parsedArgs["port"]) {
  serverSettingsJS.port = parsedArgs["port"];
}
if (parsedArgs["server-host"]) {
  serverSettingsJS.serverHost = parsedArgs["server-host"];
}
if (parsedArgs["server-root"]) {
  serverSettingsJS.serverRoot = parsedArgs["server-root"];
}
if (parsedArgs["verbose"]) {
  serverSettingsJS.verbose = parsedArgs["verbose"];
}

// TODO: Remove this export
export const VERBOSE = Boolean(serverSettingsJS.verbose);
export const SERVER_SETTINGS = ServerSettings.fromJS(serverSettingsJS);

// --- Sign of Life -------------------------------
if (START_SERVER) {
  logger.log(`Starting Turnilo v${VERSION}`);
}

// --- Location -------------------------------

let settingsStore: SettingsStore;

if (configContent) {
  const appSettings = AppSettings.fromJS(configContent, {});
  // TODO: this validation should be done via #365
  appSettings.validate();
  settingsStore = SettingsStore.create(appSettings);
} else {
  let initAppSettings = AppSettings.BLANK;

  // If a file is specified add it as a dataCube
  const fileToLoad = parsedArgs["file"];
  if (fileToLoad) {
    initAppSettings = initAppSettings.addDataCube(new DataCube({
      name: path.basename(fileToLoad, path.extname(fileToLoad)),
      clusterName: "native",
      source: fileToLoad
    }));
  }

  const url = parsedArgs.druid;
  if (url) {
    initAppSettings = initAppSettings.addCluster(new Cluster({
      name: "druid",
      url,
      sourceListScan: "auto",
      sourceListRefreshInterval: Cluster.DEFAULT_SOURCE_LIST_REFRESH_INTERVAL,
      sourceListRefreshOnLoad: Cluster.DEFAULT_SOURCE_LIST_REFRESH_ON_LOAD,
      sourceReintrospectInterval: Cluster.DEFAULT_SOURCE_REINTROSPECT_INTERVAL,
      sourceReintrospectOnLoad: Cluster.DEFAULT_SOURCE_REINTROSPECT_ON_LOAD
    }));
  }

  settingsStore = SettingsStore.create(initAppSettings);
}

export const SETTINGS_MANAGER = new SettingsManager(settingsStore, {
  logger,
  verbose: VERBOSE,
  anchorPath: configDirPath,
  initialLoadTimeout: SERVER_SETTINGS.getPageMustLoadTimeout()
});

// --- Printing -------------------------------

if (PRINT_CONFIG) {
  var withComments = Boolean(parsedArgs["with-comments"]);

  SETTINGS_MANAGER.getFreshSettings({
    timeout: 10000
  }).then(appSettings => {
    const config = appSettingsToYAML(appSettings, withComments, {
      header: true,
      version: VERSION,
      verbose: VERBOSE,
      port: SERVER_SETTINGS.getPort()
    });
    process.stdout.write(config, () => process.exit());
  }).catch((e: Error) => {
    exitWithError("There was an error generating a config: " + e.message);
  });
}
